### PR TITLE
initial working implementation of JDBC tracing via p6spy

### PR DESCRIPTION
**Pull request by ewhauser**
This PR address #201. It is ready for review, but is not ready to be merged.  Outstanding issues:

- [X] ~~p6spy needs to release a version of 3.0.0 to Maven Central (the build will fail until this is resolved)~~
- [ ] address #206 (optional) #

### CODE REVIEW

**Comment by adriancole** [comment]
Thanks for raising this. one thing to think about: should we call this brave-p6spy?

The code is centered around p6spy, which is jdbc focused, but a change to p6spy might warrant a change here (ex if an incompatible version update happens). On the other hand, folks may not recognize p6spy as a jdbc library, so naming it brave-p6spy might not help people understand what it does.

so it is a choice between

brave-jdbc
brave-p6spy
brave-jdbc-p6spy

**Comment by ewhauser** [comment]
If we think `brave-jdbc` is too general then I would probably vote for changing it to `brave-jdbc-p6spy`

**Comment by eirslett** [comment]
We could do like brave-http. That's a generic library that's then again used by actual http libraries, the dependency is like this:

application -> brave-okhttp -> brave-http -> brave-core

We could do something similar:

application -> brave-p6spy -> brave-jdbc -> brave-core

File: brave-jdbc/src/main/java/com/github/kristofa/brave/jdbc/BraveJdbcListener.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import zipkin.Constants;
+import zipkin.TraceKeys;
+
+import java.net.Inet4Address;
+import java.net.InetAddress;
+import java.net.UnknownHostException;
+import java.nio.ByteBuffer;
+import java.sql.SQLException;
+
+public final class BraveJdbcListener extends JdbcEventListener {
```
Comment (felixbarny) [inline brave-jdbc/src/main/java/com/github/kristofa/brave/jdbc/BraveJdbcListener.java]: I think you can extend [`SimpleJdbcEventListener`](https://github.com/p6spy/p6spy/blob/master/src/main/java/com/p6spy/engine/event/SimpleJdbcEventListener.java) to reduce code duplication

File: brave-jdbc/src/test/java/com/github/kristofa/brave/jdbc/BraveJdbcBenchmarks.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+ * driver.
+ */
+@Measurement(iterations = 10, time = 1)
+@Warmup(iterations = 10, time = 1)
+@Fork(1)
+@BenchmarkMode(Mode.Throughput)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+@Threads(1)
+@State(Scope.Benchmark)
+public class BraveJdbcBenchmarks {
```
Comment (adriancole) [inline brave-jdbc/src/test/java/com/github/kristofa/brave/jdbc/BraveJdbcBenchmarks.java]: thoughts about moving this into the benchmarks module?

File: brave-jdbc/src/test/java/com/github/kristofa/brave/jdbc/DerbyUtils.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package com.github.kristofa.brave.jdbc;
+
+import java.io.OutputStream;
+
+public class DerbyUtils {
+
+    //Get rid of the annoying derby.log file
```
Comment (adriancole) [inline brave-jdbc/src/test/java/com/github/kristofa/brave/jdbc/DerbyUtils.java]: heh :)

File: brave-jdbc/src/main/java/com/github/kristofa/brave/jdbc/BraveJdbcListener.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        } catch (UnknownHostException ignored) {
+        }
+        try {
+            port = Integer.parseInt(options.getPort());
+        } catch (Exception ignored) {
+        }
+        serviceName = options.getServiceName();
+    }
+
+    public static void setClientTracer(final ClientTracer tracer) {
```
Comment (adriancole) [inline brave-jdbc/src/main/java/com/github/kristofa/brave/jdbc/BraveJdbcListener.java]: I'd scrub the "final " on the params in this file.. I think most are

File: brave-jdbc/src/main/java/com/github/kristofa/brave/jdbc/P6BraveOptionsMBean.java [hunk 5]
```diff
@@ -0,0 +1,11 @@
+package com.github.kristofa.brave.jdbc;
+
+public interface P6BraveOptionsMBean {
```
Comment (adriancole) [inline brave-jdbc/src/main/java/com/github/kristofa/brave/jdbc/P6BraveOptionsMBean.java]: are these only used for endpoint? If so, add javadoc links to Endpoint#XXX

File: brave-jdbc/src/test/java/com/github/kristofa/brave/jdbc/BraveJdbcTests.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+            assertEquals(count[0], 1);
+            assertEquals(count[1], 1);
+        }, PREPARED_INSERT_SQL);
+        validateTraced(clientTracer, PREPARED_INSERT_SQL);
+    }
+
+    private void validateTraced(ClientTracer clientTracer, String sql) {
+        final InOrder order = inOrder(clientTracer);
+
+        order.verify(clientTracer).startNewSpan("query");
```
Comment (adriancole) [inline brave-jdbc/src/test/java/com/github/kristofa/brave/jdbc/BraveJdbcTests.java]: are you cool with all of these operations having the same name?

**Comment by adriancole** [comment]
we could introduce brave-jdbc (or brave-sql) to normalize how the span looks, if we ever end up having operation names not being "query" or similar. Ex. this is duplicated between mysql and here.

While I think this could be the best way for folks who cannot control libraries on top of JDBC drivers, I hesitate about calling this brave-jdbc. Ex. things like Jooq are nicer when you can control the application layer, as it doesn't require static (or otherwise) collaboration. In either case, the thing that avails you the tracing interceptor is the library, not JDBC. If one day JDBC has an interceptor, well that would be brave-jdbc.

long story short, I like `brave-p6spy` best despite it requiring people to know p6spy is a thing. I think it is probably good that they know this, since that's the thing doing the work anyway.

**Comment by adriancole** [comment]
on the code and docs itself, looks fine and ready to go once minor comments are addressed.

**Comment by adriancole** [comment]
@ewhauser you back from holiday? :D

File: brave-jdbc/src/main/java/com/github/kristofa/brave/jdbc/BraveJdbcListener.java (hunk 1 above)
Comment (ewhauser) [inline brave-jdbc/src/main/java/com/github/kristofa/brave/jdbc/BraveJdbcListener.java]: I did try this, but that interface doesn't appear to work correctly for insert statements.  The problem appears to be that the `String` of SQL that is passed in when an insert is executed seems to get lost in `SimpleJdbcEventListener` as it is not set on the `StatementInformation`

File: brave-jdbc/src/test/java/com/github/kristofa/brave/jdbc/BraveJdbcBenchmarks.java (hunk 2 above)
Comment (ewhauser) [inline brave-jdbc/src/test/java/com/github/kristofa/brave/jdbc/BraveJdbcBenchmarks.java]: This depends on a bunch of other classes in the test project, so not sure it's worth it to try to bundle those up as shareable so that this class could be in the benchmarks module.

File: brave-jdbc/src/test/java/com/github/kristofa/brave/jdbc/BraveJdbcTests.java (hunk 6 above)
Comment (ewhauser) [inline brave-jdbc/src/test/java/com/github/kristofa/brave/jdbc/BraveJdbcTests.java]: not sure I follow

**Comment by ewhauser** [comment]
Comments addressed.  Awaiting resolution on #206 (could be addressed separately)

**Comment by adriancole** [comment]
Looks great! thanks for wrapping this up.
